goben/prober.go

```diff
@@ -0,0 +1,289 @@
+func (p *Prober) packetToRecv(pktbuf []byte) (net.IP, *icmp.Message, error) {
+	n, sender, err := p.conn.ReadFrom(pktbuf)
+	if err != nil {
+		return nil, nil, err
+	}
+	// get the sender's IP address
+	// Since sender is an interface net.Addr, we have to cast it down to net.IPAddr/net.UDPAddr type to get the IP
+	var senderIP net.IP
+	switch sender := sender.(type) {
+	case *net.IPAddr:
+		senderIP = sender.IP
+	case *net.UDPAddr:
+		senderIP = sender.IP
+	}
+
+	msg, err := icmp.ParseMessage(protocolICMP, pktbuf[:n])
+	if err != nil {
+		return nil, nil, err
+	}
+	return senderIP, msg, nil
+}
+
+// send ICMP message
+// We sent the icmp packets synchronously
+func (p *Prober) send(runID uint16, morePkts chan bool) {
+	seq := runID & uint16(0xff00)
+	for i := 0; i < p.config.pktsPerProbe; i++ {
+		for _, target := range p.config.targets {
+			log.Printf("Request to=%s id=%d seq=%d", target, runID, seq)
+			if _, err := p.conn.WriteTo(p.packetToSend(runID, seq), parseIP(target)); err != nil {
+				log.Println(err.Error())
+				continue
+			}
+			fmt.Println("SEND A ICMP PACKET")
+			morePkts <- true
+		}
+		seq++
+		time.Sleep(p.config.pktInterval)
+	}
+	log.Printf("%s: Done sending packets, closing the tracker.", p.config.source)
+	close(morePkts)
+}
+
+// receive ICMP message and compute the measurements
+func (p *Prober) recv(runID uint16, morePkts chan bool) {
+	// keep track if the packet arrived has been received before
+	received := make(map[string]bool)
+	// a counter to make sure we read all packets arrived (including the outstanding
+	// ones after the sender has closed the connection)
+	outstandingPkts := 0
+	// the byte stream buffer
+	pktbuf := make([]byte, 1500)
+
+	for {
+		if outstandingPkts == 0 {
+			if _, ok := <-morePkts; ok {
+				outstandingPkts++
+			} else {
+				return
+			}
+		}
+		senderIP, msg, err := p.packetToRecv(pktbuf)
+		if err != nil {
+			p.fileLogger.Printf("Unmarshalling icmp message Error: %s\n", err.Error())
+			if neterr, ok := err.(*net.OpError); ok && neterr.Timeout() {
+				return
+			}
+		}
+		target := senderIP.String()
+		echoMsg, ok := msg.Body.(*icmp.Echo)
+		if !ok {
+			p.fileLogger.Println("Got wrong packet in ICMP echo reply.") // should never happen
+			continue
+		}
+
+		// get rtt
+		rtt := time.Since(bytesToTime(echoMsg.Data))
+
+		// check if this packet belong to this run
+		if !matchPacket(runID, echoMsg.ID, echoMsg.Seq) {
+			p.fileLogger.Printf(
+				"Reply from=%s id=%d seq=%d rtt=%s Unmatched packet, probably from the last probe run.\n",
+				target, echoMsg.ID, echoMsg.Seq, rtt)
+			continue
+		}
+
+		// check if we have seen this packet before
+		pktID := fmt.Sprintf("%s_%d", target, echoMsg.Seq)
+		if received[pktID] {
+			p.fileLogger.Printf("Duplicate reply from=%s id=%d seq=%d rtt=%s\n", target, echoMsg.ID, echoMsg.Seq, rtt)
+			continue
+		}
+
+		// record the rrt
+		// todo: make this an option that the user can choose to verbosely logging in the console and/or dump to a file
```

Comment:
It should exactly work like the csv option in the original goben traffic generator.

Reply:
done